pkg/async/worker_test.go

```diff
@@ -283,3 +283,78 @@ func TestWorkerReceiveAndWorkBlocksUntilContextCanceled(t *testing.T) {
 	err := w.receiveAndWork(ctx, getDisposableQueueName())
 	assert.Equal(t, ctx.Err(), err)
 }
+
+func TestWorkerDoesNotMoveTaskEarly(t *testing.T) {
+
+	mainActiveWorkQueueName := getDisposableQueueName()
+	mainDelayedWorkQueueName := getDisposableQueueName()
+
+	task := model.NewDelayedTask("jobName",
+		map[string]string{
+			"stepName":   "step",
+			"instanceID": "provision",
+		},
+		time.Minute*10,
+	)
+	taskJSON, _ := task.ToJSON()
+	redisClient.LPush(mainDelayedWorkQueueName, taskJSON)
+	w := newWorker(redisClient).(*worker)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*10)
+	defer cancel()
+	_ = w.handleDelayedTasks(
```

Comment:
I think there's a more "focused" way to test this. This function call invokes the _loop_ that listens to the main delayed work queue and spawns new goroutines that launch timers and do the eventual move. If what we want to test is just that tasks aren't moved prematurely, we can test the `handleDelayedTask()` (singular / non-loop) function directly. Same applies to the other new test in this commit.

Reply:
We can write separate tests for the loop. They'll just verify that `handleDelayedTask()` is invoked once per delayed task.

Reply:
^ And I will happily contribute that loop test.